### PR TITLE
Add support for CVSSv4

### PR DESCRIPTION
**Pull request by jobiewinserapck**
-Added class CVSS4
-Based the CVSS4 class on the js implementation found here: https://github.com/RedHatProductSecurity/cvss-v4-calculator -Added new tests for CVSS4
-Added new constants and exceptions for CVSS4
-Added json schema for CVSS4

### CODE REVIEW

**Comment by skontar** [comment]
This looks really good. Thank you!

**Comment by skontar** [comment]
@jobiewinserapck I presume you used the js implementation to generate test vectors?

**Comment by jobiewinserapck** [comment]
Yes, I generated the vectors using the js implementation. I could Maybe get some screenshots tomorrow of how I did so. If thats helpful

**Comment by skontar** [comment]
@jobiewinserapck can you rebase with master? I think that should re-run the CI with latest setup.

**Comment by jobiewinserapck** [comment]
Sure thing I'll do that ASAP tomorrow morning

File: cvss/cvss4.py [hunk 1]
```diff
@@ -0,0 +1,643 @@
...
+
+    def clean_vector(self, output_prefix=True):
+        """
+        Returns vector without optional metrics marked as X and in preferred order.
+
+        Args:
+            output_prefix (bool): defines if CVSS vector should be printed with prefix
+
+        Returns:
+            (str): cleaned CVSS3 with metrics in correct order
```
Comment (skontar) [inline cvss/cvss4.py:560]: Should be 4.

File: cvss/cvss4.py [hunk 2]
```diff
@@ -0,0 +1,643 @@
...
+        """
+        vector = []
+        for metric in METRICS_ABBREVIATIONS:
+            if metric in self.original_metrics:
+                value = self.original_metrics[metric]
+                if value != "X":
+                    vector.append("{0}:{1}".format(metric, value))
+        if output_prefix:
+            prefix = "CVSS:4/"
```
Comment (skontar) [inline cvss/cvss4.py:569]: CVSS:4.0?

File: cvss/cvss4.py [hunk 3]
```diff
@@ -0,0 +1,643 @@
...
+        self.add_missing_optional()
+        self.compute_base_score()
+        self.compute_severity()
+
+    def check_mandatory(self):
+        """
+        Checks if mandatory fields are in CVSS3 vector.
+
+        Raises:
+            CVSS3MandatoryError: if mandatory metric is missing in the vector
```
Comment (skontar) [inline cvss/cvss4.py:97]: CVSS4MandatoryError

File: cvss/cvss4.py [hunk 4]
```diff
@@ -0,0 +1,643 @@
...
+
+        self.parse_vector()
+        self.check_mandatory()
+        self.add_missing_optional()
+        self.compute_base_score()
+        self.compute_severity()
+
+    def check_mandatory(self):
+        """
+        Checks if mandatory fields are in CVSS3 vector.
```
Comment (skontar) [inline cvss/cvss4.py:94]: CVSS4

File: cvss/cvss4.py [hunk 5]
```diff
@@ -0,0 +1,643 @@
...
+
+        Raises:
+            CVSS4MalformedError: if vector is not in expected format
+        """
+        if self.vector == "":
+            raise CVSS4MalformedError("Malformed CVSS4 vector, vector is empty")
+
+        if self.vector.endswith("/"):
+            raise CVSS4MalformedError('Malformed CVSS4 vector, trailing "/"')
+        try:
```
Comment (skontar) [inline cvss/cvss4.py:156]: Similarly as on 3, insert between 155 and 156:
```python
        # Handle 'CVSS:4.x' in the beginning of vector and split vector
        if not self.vector.startswith('CVSS:4.0/'):
            raise CVSS4MalformedError('Malformed CVSS4 vector "{0}" is missing mandatory prefix '
                                      'or uses unsupported CVSS version'.format(self.vector))
```

**Review by skontar — CHANGES_REQUESTED** [review]

**Comment by skontar** [comment]
@mprpic can you please review?

**Comment by jobiewinserapck** [comment]
I'ved added those improvements, thanks for that @skontar!

**Comment by marco-silva0000** [comment]
just commenting for support, gogogo

**Review by skontar — APPROVED** [review]

**Review by mprpic — COMMENTED** [review]
Can you also please add import statements into the `cvss/__init__.py` file?

```
from .cvss4 import CVSS4
from .exceptions import CVSS4Error
```

**Comment by jobiewinserapck** [comment]
I've pushed up a commit addressing the issues raised by @mprpic.

I also added some additional tests for malformed cvss strings

**Comment by skontar** [comment]
@mprpic please re-review.

**Review by mprpic — APPROVED** [review]
LGTM!

**Comment by AdrianVollmer** [comment]
I don't think `round_away_from_zero` is correct:

```console
$ python << EOF
from decimal import ROUND_HALF_UP
from decimal import Decimal as D
def round_away_from_zero(x, dp=1):
    return float(D(x).quantize(D("0." + "0" * dp), rounding=ROUND_HALF_UP))

for i in range(6):
    print(round_away_from_zero(i + 0.05))
EOF
0.1
1.1
2.0
3.0
4.0
5.0
```

It's inconsistent.

I would have said we should just do `round(x*10)/10` like the JavaScript implementation, but `Math.round(0.5)` is 1 in JS and `round(0.5)` is 0 in Python3 and 1 in Python2.  Python3 follows [IEEE 754](https://en.wikipedia.org/w/index.php?title=IEEE_754#Rounding_rules).

The specification document is not clear on which rounding convention should be used:

>     This score is rounded to one decimal place.

**Comment by skontar** [comment]
Correct `.quantize` call should be able to resolve that difference. I do not recall what we did for v3.1 but we were able to get exactly the same results between Python and Javascript.

**Comment by skontar** [comment]
@AdrianVollmer if the input is `Decimal` then `round_away_from_zero` works correctly I think.
```
>>> float((D(2) + D(5) / D(100)).quantize(D("0.0"), rounding=ROUND_HALF_UP))
2.1
```

**Comment by AdrianVollmer** [comment]
Okay, but this function is only used once here and is passed a float:

https://github.com/RedHatProductSecurity/cvss/blob/c5dd7e2f0bcda1b32d509cce59a6db7c94bbec64/cvss/cvss4.py#L558

**Comment by AdrianVollmer** [comment]
I think this function will suit this particular use case:

```python
def round_away_from_zero(value):
    # Only works correctly for positive values
    return int(value * 10 + 0.5) / 10.0
```

Since we only need to round values between 0.0 and 10.0, it should be fine.

```console
$ python << EOF
for i in range(10):
 print(int((i+0.05)*10 + 0.5)/10)
EOF
0.1
1.1
2.1
3.1
4.1
5.1
6.1
7.1
8.1
9.1
```

**Pull request merged**